### PR TITLE
New COSE message implementation

### PR DESCRIPTION
**Pull request by vikram-gaur**
Added new implementations for COSE messages (Encrypt, Mac, Sign)

### CODE REVIEW

File: java/com/google/cose/CoseMessage.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import co.nstant.in.cbor.model.DataItem;
+import co.nstant.in.cbor.model.Map;
+import com.google.cose.utils.CborUtils;
+
+/**
+ * Implements the base class for COSE Message structure to be implemented for other message types.
+ */
+public abstract class CoseMessage {
+  private final byte[] protectedHeaderBytes;
```
Comment (MaxBires) [inline java/com/google/cose/CoseMessage.java]: Still reading further, but I think it would generally be easier from a use perspective for the library caller to provide a Map and let this library handle serializing it under the hood. I'm generally in favor of passing structured objects through APIs when possible, as opposed to bytes. E.g. I foresee users either bstr encoding the protectedHeaderBytes preemptively when they shouldn't, or vice versa.

Comment (MaxBires) [reply java/com/google/cose/CoseMessage.java]: Actually after further discussion with Paul, just providing two methods for adding protectedHeaders, either a byte array or a Map, should suffice. The latter would immediately canonically serialize the Map into the protectedHeaderBytes field.

Comment (vikram-gaur) [reply java/com/google/cose/CoseMessage.java]: My only concern was that providing both the methods may cause someone to use both the methods and providing incompatible definitions. Although we can test for that, byte array would make sure that user is providing the actual values they want. But I see your point on preemptive conversion as well.

Comment (vikram-gaur) [reply java/com/google/cose/CoseMessage.java]: Added protectedHeader map support

File: java/com/google/cose/Recipient.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ */
+public class Recipient extends CoseMessage {
+  private final byte[] ciphertext;
+  private final List<Recipient> recipients;
+
+  private Recipient(byte[] protectedHeaderBytes, Map unprotectedHeaders, byte[] ciphertext,
+      List<Recipient> recipients) {
+    super(protectedHeaderBytes, unprotectedHeaders);
+    this.ciphertext = ciphertext;
+    this.recipients = recipients;
```
Comment (sethmoo) [inline java/com/google/cose/Recipient.java]: Suggest ensuring `this.recipients` is never null, so callers don't have to worry about getting a null value when calling `getRecipients`. e.g. `this.recipients = recipients == null ? new List<>() : recipients`

File: java/com/google/cose/SignMessage.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      this.unprotectedHeaders = unprotectedHeaders;
+      return this;
+    }
+
+    public Builder withMessage(byte[] message) {
+      this.message = message;
+      return this;
+    }
+
+    public Builder withSignatures(List<Signature> signatures) {
```
Comment (sethmoo) [inline java/com/google/cose/SignMessage.java]: I think a more idiomatic way of doing this would be to have an `addSignature(Signature signature)` and each time it's called it extends the list with the new signature.

File: java/com/google/cose/exceptions/CoseException.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.google.cose.exceptions;
+
+/**
+ * This class is used for throwing any exceptions that may occur during parsing and serialization of
+ * byte data into their respective COSE components.
+ */
+public class CoseException extends RuntimeException {
```
Comment (sethmoo) [inline java/com/google/cose/exceptions/CoseException.java]: Are we sure this is a `RuntimeException`? It seems like we might want this checked, as we're going to throw on bad data. If callers aren't made explicitly aware of the exception, then they may not catch. Bad data crashing our services seems unwanted.

File: java/com/google/cose/exceptions/CoseException.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  public CoseException(final String message) {
+    super(message);
+  }
+
+  public CoseException(final String message, final Exception ex) {
```
Comment (sethmoo) [inline java/com/google/cose/exceptions/CoseException.java]: nit: I think the second parameter should be `Throwable`, not `Exception`.

File: java/com/google/cose/structures/EncryptStructure.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Encodes the Enc_Structure as mentioned in COSE RFC section 5.3
+ */
+public class EncryptStructure {
+  private final String context;
+  private final Map protectedHeaders;
+  private final byte[] externalAad;
+
+  public EncryptStructure(String context, Map headers, byte[] externalAad) {
+    this.context = context;
```
Comment (sethmoo) [inline java/com/google/cose/structures/EncryptStructure.java]: Can you make context an enumeration? IIUC, there's a very limited set of allowed values, and making an enumeration will help prevent silly mistakes like typos.

Comment (sethmoo) [reply java/com/google/cose/structures/EncryptStructure.java]: The same comment applies to all the structure types. I think their context values are all very limited.

File: java/com/google/cose/utils/Algorithm.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+  private final int algorithmId;
+  private final String instance;
+
+  Algorithm(int algorithmId, String instance) {
+    this.algorithmId = algorithmId;
+    this.instance = instance;
+  }
+
+  public String getInstance() {
```
Comment (sethmoo) [inline java/com/google/cose/utils/Algorithm.java]: Suggest `javaAlgorithm` and 'coseAlgorithm` instead for the fields.

I found `Instance` confusing because that implies it's returning an instance of some `Algorithm`, but it's a parameter to be passed as the `algorithm` for `javax.crypto`.

Since both COSE and Java refer to these as "algorithm", then I think we should be consistent with their naming.

File: java/com/google/cose/utils/CborUtils.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    return (ByteString) dataItem;
+  }
+
+  /**
+   * Returns DataItem from a cbor map based on Integer index.
+   * @param cborMap map that has the information.
+   * @param index integer index to be used as key in the map.
+   * @return value in the map corresponding to key
+   */
+  public static DataItem getValueFromMap(final Map cborMap, final int index) {
```
Comment (sethmoo) [inline java/com/google/cose/utils/CborUtils.java]: Unused?

File: java/com/google/cose/utils/CborUtils.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+import co.nstant.in.cbor.model.UnsignedInteger;
+import com.google.cose.exceptions.CoseException;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.util.List;
+
+/**
+ * This class contains utility functions for converting bytes to CBOR objects.
+ */
+public class CborUtils {
```
Comment (sethmoo) [inline java/com/google/cose/utils/CborUtils.java]: Suggest splitting this into two classes: one for public utilities, one for protected. AFAICT, some of these methods are only ever useful internally to the package.

File: java/com/google/cose/utils/Headers.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
+package com.google.cose.utils;
+
+/**
+ * This class contains all the header keys that we are going to be using in COSE implementation.
+ */
+public class Headers {
```
Comment (sethmoo) [inline java/com/google/cose/utils/Headers.java]: Nit: these aren't just header keys. These are simply COSE ids defined by IANA for various purposes.

File: test/com/google/cose/Encrypt0MessageTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+public class Encrypt0MessageTest {
+  @Test
+  public void testDeserialize() {
+    Encrypt0Message message = Encrypt0Message.deserialize(TestUtilities.hexStringToByteArray(
+      "8343A10101A1054C02D1F7E6F26C43D4868D87CE582460973A94BB2898009EE52ECFD9AB1DD25867374B162E2C0"
+          + "3568B41F57C3CC16F9166250B"
+    ));
+    Assert.assertEquals("60973A94BB2898009EE52ECFD9AB1DD25867374B162E2C03568B41F57C3CC16F9166250B",
+        TestUtilities.bytesToHexString(message.getCiphertext()));
+    Assert.assertEquals("A10101",
```
Comment (sethmoo) [inline test/com/google/cose/Encrypt0MessageTest.java:25]: I assume this is actually a `Map` now in the latest revision (per Max's feedback above)?

File: test/com/google/cose/SignMessageTest.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    Assert.assertEquals(1, s.getUnprotectedHeaders().getKeys().size());
+    Assert.assertEquals(new ByteString(TestUtilities.hexStringToByteArray("3131")),
+        s.getUnprotectedHeaders().get(new UnsignedInteger(4)));
+    Assert.assertEquals("E2AEAFD40D69D19DFE6E52077C5D7FF4E408282CBEFB5D06CBF414AF2E19D982AC45AC98B"
+        + "8544C908B4507DE1E90B717C3D34816FE926A2B98F53AFD2FA0F30A",
+        TestUtilities.bytesToHexString(s.getSignature()));
+  }
+
+  @Test
+  public void testDeserialize2() {
```
Comment (sethmoo) [inline test/com/google/cose/SignMessageTest.java:38]: Please give these slightly more descriptive names. Maybe `testDeserializeWithEmptyProtectedHeaderString` and `testDeserializeWithEmptyProtectedHeaderMap` or something to indicate the difference. (yeah, those are long names, I'm open to better options).

File: test/com/google/cose/TestUtilities.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package com.google.cose;
+
+public class TestUtilities {
+  public static final String CONTENT = "This is the content.";
```
Comment (sethmoo) [inline test/com/google/cose/TestUtilities.java:4]: nit: I think `"This is the content."` as a literal is actually more readable. Up to you if you want to change it.

File: test/com/google/cose/structure/SignStructureTest.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+import com.google.cose.structures.SignStructure;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class SignStructureTest {
+  @Test
+  public void testHappyCase() {
```
Comment (sethmoo) [inline test/com/google/cose/structure/SignStructureTest.java]: Shouldn't we also test sad cases? (this is a general comment. feel free to add more tests later for negative paths)

**Review by sethmoo — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+  <scm>
+    <connection>scm:git:https://github.com/android/cose-lib.git</connection>
+    <developerConnection>scm:git:git@github.com:android/cose-lib.git</developerConnection>
+    <url>https://github.com/android/cose-lib</url>
+  </scm>
+  <developers>
+    <developer>
+      <id>vikram-gaur</id>
+      <name>Vikram Gaur</name>
+      <email>vikramgaur@google.com</email>
```
Comment (sethmoo) [inline pom.xml:53]: Should this be a group that external folks can email?

File: pom.xml [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+        <artifactId>j2objc-annotations</artifactId>
+        <version>1.3</version>
+      </dependency>
+      <dependency>
+        <groupId>junit</groupId>
+        <artifactId>junit</artifactId>
+        <version>4.13.2</version>
+        <scope>test</scope>
+      </dependency>
+      <dependency>
```
Comment (sethmoo) [inline pom.xml:203]: Are we actually using mocks?

**Review by sethmoo — APPROVED** [review]

**Review by sethmoo — APPROVED** [review]

File: pom.xml [hunk 17]
```diff
@@ -137,28 +134,6 @@
             </execution>
           </executions>
         </plugin>
-        <plugin>
```
Comment (sethmoo) [inline pom.xml]: The animal sniffer plugin looks super helpful. Maybe file a bug to use it later for API stability?

File: pom.xml [hunk 18]
```diff
@@ -219,74 +194,12 @@
   </build>
   <dependencyManagement>
     <dependencies>
-      <dependency>
-        <groupId>com.google.code.findbugs</groupId>
```
Comment (sethmoo) [inline pom.xml]: Some of these checkers are great (like errorprone). We should keep them, IMO.

**Review by sethmoo — CHANGES_REQUESTED** [review]

**Review by sethmoo — CHANGES_REQUESTED** [review]
nit: You didn't change header values. You changed the header labels with constants. The description was kind of confusing to me.

Also, it's kind of odd that you just slipped in new testcases at the same time. That really should be a separate commit.

**Review by sethmoo — APPROVED** [review]
A few notes:
1. That's a lot of copy pasta!
2. Why not make `protectedHeaderBytes` `protectedHeaders` and treat it as a map object? Anyone who is using `getProtectedHeaders` almost certainly doesn't want bytes (right?). They want to look at the headers.
3. What is the use case for creating an object with pre-encoded bytes?

File: src/com/google/cose/Encrypt0Message.java [hunk 19]
```diff
@@ -37,6 +37,18 @@ public Builder withProtectedHeaderBytes(byte[] protectedHeaderBytes) {
       return this;
     }
 
+    public Builder withProtectedHeaders(Map protectedHeaders) {
+      if (protectedHeaderBytes != null) {
+        throw new CoseException("Cannot use both withProtectedHeaderBytes and withProtectedHeaders");
```
Comment (sethmoo) [inline src/com/google/cose/Encrypt0Message.java:42]: This string is misleading. As it stands, we cannot differentiate if someone called this method twice _or_ `withProtectedHeaderBytes` followed by `withProtectedHeaders`.

**Review by sethmoo — CHANGES_REQUESTED** [review]
Shoot, I didn't mean to approve :) I think this is how I un-approve on github?

File: java/com/google/cose/CoseMessage.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
+package com.google.cose;
```
[the remark this answers is not in the document]
Comment (vikram-gaur) [reply java/com/google/cose/CoseMessage.java]: added

File: src/com/google/cose/CoseMessage.java [hunk 21]
```diff
@@ -1,3 +1,19 @@
+/*
+ * Copyright 2020 Google LLC
```
Comment (sethmoo) [inline src/com/google/cose/CoseMessage.java:2]: Wrong year :)

File: java/com/google/cose/Encrypt0Message.java [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+        throw new CoseException("Error while decoding Encrypt0Message. Expected 3 items,"
+            + "received " + messageArray.size());
+      }
+      Encrypt0Message.Builder builder = new Encrypt0Message.Builder();
+      builder.withProtectedHeaderBytes(CborUtils.asByteString(messageArray.get(0)).getBytes());
+      builder.withUnprotectedHeaders(CborUtils.asMap(messageArray.get(1)));
+      builder.withCiphertext(CborUtils.asByteString(messageArray.get(2)).getBytes());
+      return builder.build();
+    } catch (CborException ex) {
+      throw new CoseException("Error while decoding Encrypt0Message", ex);
```
[the remark this answers is not in the document]
Comment (vikram-gaur) [reply java/com/google/cose/Encrypt0Message.java]: I wanted to keep a single throwable exception throughout the project so that developers can simply handle a single instance of exception instead of dealing with multiple exceptions which may stem from the same root cause.

File: java/com/google/cose/Recipient.java (hunk 2 above)
Comment (vikram-gaur) [reply java/com/google/cose/Recipient.java]: Recipients field in Recipient class is a nullable field actually since the cose spec mentions that it may or may not be present for this structure.

File: java/com/google/cose/SignMessage.java (hunk 3 above)
Comment (vikram-gaur) [reply java/com/google/cose/SignMessage.java]: or withSignatures(Signature...signatures) which can take multiple parameters and just extends the signature list.

File: java/com/google/cose/utils/CborUtils.java (hunk 8 above)
Comment (vikram-gaur) [reply java/com/google/cose/utils/CborUtils.java]: yeah. This is being used for Key constructors. I can remove this for now and add it again later.

File: pom.xml (hunk 18 above)
Comment (vikram-gaur) [reply pom.xml]: I can add them at a later stage.

File: java/com/google/cose/Encrypt0Message.java [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+
+  static class Builder {
+    private byte[] protectedHeaderBytes;
+    private Map unprotectedHeaders;
+    private byte[] ciphertext;
+    public Encrypt0Message build() {
+      if ((protectedHeaderBytes != null) && (unprotectedHeaders != null) && (ciphertext != null)) {
+        return new Encrypt0Message(protectedHeaderBytes, unprotectedHeaders, ciphertext);
+      } else {
+        throw new CoseException("Some fields are missing.");
```
[the remark this answers is not in the document]
Comment (vikram-gaur) [reply java/com/google/cose/Encrypt0Message.java]: I want to use a single pattern, hence stuck to Builder even when all the fields are required.

File: java/com/google/cose/exceptions/CoseException.java (hunk 4 above)
Comment (vikram-gaur) [reply java/com/google/cose/exceptions/CoseException.java]: Updated this to a checked exception now.

File: java/com/google/cose/exceptions/CoseException.java (hunk 5 above)
Comment (vikram-gaur) [reply java/com/google/cose/exceptions/CoseException.java]: either works fine. But I have updated the code to make use of throwable now. However, I am not making explicit use of either for now.

File: java/com/google/cose/structures/EncryptStructure.java (hunk 6 above)
Comment (vikram-gaur) [reply java/com/google/cose/structures/EncryptStructure.java]: Converted in all structure types.

File: java/com/google/cose/utils/Algorithm.java (hunk 7 above)
Comment (vikram-gaur) [reply java/com/google/cose/utils/Algorithm.java]: Done.

File: test/com/google/cose/Encrypt0MessageTest.java (hunk 11 above)
Comment (vikram-gaur) [reply test/com/google/cose/Encrypt0MessageTest.java:43]: yes. But we had a conversation about keeping both header bytes as well as headers for now.

File: test/com/google/cose/SignMessageTest.java (hunk 12 above)
Comment (vikram-gaur) [reply test/com/google/cose/SignMessageTest.java]: done

File: java/com/google/cose/Encrypt0Message.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (sethmoo) [reply java/com/google/cose/Encrypt0Message.java]: OH! I misread this the first time. I thought you were catching a `CoseException` and throwing a new one.

File: pom.xml (hunk 16 above)
Comment (sethmoo) [reply pom.xml:203]: I see this is removed elsewhere.

File: src/com/google/cose/Encrypt0Message.java (hunk 19 above)
Comment (sethmoo) [reply src/com/google/cose/Encrypt0Message.java:58]: Maybe say "protected headers may only be added once"?

File: test/com/google/cose/Encrypt0MessageTest.java (hunk 11 above)
Comment (sethmoo) [reply test/com/google/cose/Encrypt0MessageTest.java:43]: Ah ok. SGTM.

**Review by sethmoo — APPROVED** [review]
